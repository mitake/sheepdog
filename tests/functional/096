#! /bin/bash

# Test iSCSI multipath functionality

. ./common

_need_to_be_root

which $TGTD > /dev/null || _notrun "Require tgtd but it's not installed"
which $TGTADM > /dev/null || _notrun "Require tgtadm but it's not installed"
which $ISCSID > /dev/null || _notrun "Require iscsid but it's not installed"
which $ISCSIADM > /dev/null || _notrun "Require iscsiadm but it's not installed"
which $MULTIPATHD > /dev/null || _notrun "Require multipathd but it's not installed"
which $SCSI_ID > /dev/null || _notrun "Require scsi_id but it's not installed"

$ISCSIADM -m node --logout 2> /dev/null
pkill -9 $ISCSID > /dev/null
pkill -9 $ISCSIADM > /dev/null
pkill -9 $MULTIPATHD > /dev/null
pkill -9 $TGTD > /dev/null
pkill -9 $TGTADM > /dev/null

ORIG_DEVFILES=095_orig_devfiles
LOGIN_DEVFILES=095_login_devfiles
DIFF_DEVFILES=095_diff_devfiles

/bin/ls /dev/sd* > $ORIG_DEVFILES

for i in `seq 0 2`; do
      _start_sheep $i
done

_wait_for_sheep 3
_cluster_format

$DOG vdi create test 128M

declare -a TGTD_PIDS

_setup_tgtd()
# $1: control port
# $2: iscsi portal
# $3: VDI name for backing store
{
    $TGTD -t 1 -C $1 --iscsi portal=$2
    $TGTADM -C $1 --lld iscsi --mode target --op new --tid 1 --targetname iqn.2014-06.org.sheepdog-project
    $TGTADM -C $1 --mode logicalunit --op new --tid 1 --lun 1 --bstype sheepdog --backing-store $3
    $TGTADM -C $1 --mode target --op bind --tid 1 --initiator-address ALL
}
PORTAL1=127.0.0.1:3260
PORTAL2=127.0.0.1:3261

_setup_tgtd 0 $PORTAL1 unix:$STORE/0/sock:test
_setup_tgtd 1 $PORTAL2 unix:$STORE/1/sock:test

$DOG vdi lock list

exit

ISCSID -c `pwd`/iscsid.conf

$ISCSIADM -m discovery -t sendtargets -p $PORTAL1
$ISCSIADM -m discovery -t sendtargets -p $PORTAL2

$ISCSIADM -m node --login
sleep 15

/bin/ls /dev/sd* > $LOGIN_DEVFILES

comm -3 $LOGIN_DEVFILES $ORIG_DEVFILES > $DIFF_DEVFILES
f [[ "2 $DIFF_DEVFILES" != `wc -l $DIFF_DEVFILES` ]]
then
    _notrun "Device files were not created correctly"
fi

declare -a SCSI_IDS
IDX=0
for line in `cat $DIFF_DEVFILES`; do
    SCSI_IDS[$IDX]=`$SCSI_ID --whitelisted --device=$line`
    IDX=`expr $IDX + 1`
done

if [ ${SCSI_IDS[0]} != ${SCSI_IDS[1]} ]
then
    _notrun "SCSI IDs are different: ${SCSI_IDS[0]}, ${SCSI_IDS[1]}"
fi

SCSI_ID=${SCSI_IDS[0]}

MULTIPATH_CONF=`pwd`/multipath.conf

cat <<EOF > $MULTIPATH_CONF
multipaths {
        multipath {
                wwid                    $MULTIPATH_CONF
                alias                   sheepdog-iscsi
                path_grouping_policy    multibus
                path_checker            readsector0
                path_selector           "round-robin 0"
                failback                manual
                rr_weight               priorities
                no_path_retry           5
        }
}
EOF

if [ ! -b /dev/mapper/sheepdog-iscsi ]
then
    _notrun "device file doesn't exist"
fi

file /dev/mapper/sheepdog-iscsi

if [ ! -d mnt ]
then
    mkdir mnt
else
    umount mnt 2> /dev/null
fi

mkfs.ext4 /dev/mapper/sheepdog-iscsi
mount /dev/mapper/sheepdog-iscsi mnt

dd if=/dev/urandom of=mnt/rand bs=1M count=64 of=direct iflag=fullblock
BEFORE=`md5sum mnt/rand`

# kill the default tgtd with control port 0
for pid in `pgrep -f "tgtd -C 0"`; do
    kill -9 $pid 2> /dev/null
done
sleep 20			# wait fallback
echo 3 > /proc/sys/vm/drop_caches # drop page caches
AFTER=`md5sum mnt/rand`

if [[ $BEFORE = $AFTER ]]
then
    echo "checksum matched, success"
else
    echo "checksum not matched, fail"
fi

umount mnt
