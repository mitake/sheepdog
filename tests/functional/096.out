QA output created by 095
using backend plain store
Try `iscsiadm --help' for more information.
Logging in to [iface: default, target: iqn.2014-06.org.sheepdog-project, portal: 127.0.0.1,3260] (multiple)
Logging in to [iface: default, target: iqn.2014-06.org.sheepdog-project, portal: 127.0.0.1,3261] (multiple)
Login to [iface: default, target: iqn.2014-06.org.sheepdog-project, portal: 127.0.0.1,3260] successful.
Login to [iface: default, target: iqn.2014-06.org.sheepdog-project, portal: 127.0.0.1,3261] successful.
/dev/mapper/sheepdog-iscsi: block special
